test/testdata/infer/attached_class_intrinsic.rb

```diff
@@ -31,19 +31,25 @@ def self.bind_example(&blk); end
 
   sig {params(blk: T.proc.bind(T.attached_class).void).void}
   def self.bind_attached(&blk); end
+
+  sig {params(blk: T.proc.bind(T.self_type).void).void}
+  def self.bind_self(&blk); end
 end
 
 B.bind_example do
+  T.reveal_type(self) # error: `T.class_of(Example)`
   example = new
-  T.reveal_type(example) # error: `T.attached_class (of Example)`
+  T.reveal_type(example) # error: `Example`
```

Comment:
Despite banning anything other than "simple class names," we were still allowing generic classes to sneak in here because every class singleton class is generic.

Which means that we actually do have a change in behavior here, but it's one that I actually think is more correct, so I'm inclined to keep it (though I will have to test on Stripe's codebase).